### ISSUE TITLE
Probes are not excluded from overload manager actions

### ISSUE THREAD

**Issue by caoyukun0430** [istio/istio#41859]
**Describe the feature request**
We are recently testing overload manager with istio using custom bootstrap. Our scenario is to overload the ingressgateway heap_size configured via sending requests from large number of concurrent connections, then we see stop_accepting_connections/requests actions triggered and new requests will fail.
But we found stop_accepting_connections(also stop_accepting_requests) stops the internal probes like readiness probe, which leads to ingressgateway pod stuck in state 0/1 running, which will causes issue for us.

**Describe alternatives you've considered**
We think it might not  be an approperiate way to use /healthz/ready for readiness probe under overload manager condition, we expect if istio can exclude the probes  from stop_accepting_* actions so that overload actions stop accepting requests (or connections) should not make the probes fail. Is there anything/idea istio can do to improve the probes?

related issue also raised https://github.com/envoyproxy/envoy/issues/23843

**Affected product area (please put an X in all that apply)**

[ ] Ambient
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Comment by howardjohn** [comment]
This would need envoy side changes to bypass certain listeners/routes/etc from the overload action.

Alternative, you could change the probing port to 15020 *I think* (haven't checked) which ends up bypassing envoy for readiness.

**Comment by caoyukun0430** [comment]
Hi @howardjohn thanks for the reply! I changed the readiness probe port to 15020(http://:15020/healthz/ready), so now indeed when overload action triggered, readiness probe will not fail. I did some tests and have some questions related:
1. We did some test to cross-check 15021 and 15020 difference, we found out probe request to 15020 seems to reach only pilot-agent, while 15021 reaches envoy. In a non-overload situation, we tried to **first paused/stopped envoy process using "kill -STOP PID"** and then send requests. **We found out as below request to 15000 and 15021 failed(envoy stopped) but to 15020 still succeeded**. Compared to the overload situation, where request to 15000 admin interface and 15020 succeed, but **only 15021 failed**. We would like to ask how does the request flow differ between 15020 and 15021?
```
kubectl exec -ti pod/istio-ingressgateway-7d7674c9cc-92vft -n istio  -- bash
bash-4.4$ curl http://localhost:15000/ready -v
*   Trying 127.0.0.1:15000...
* Connected to localhost (127.0.0.1) port 15000 (#0)
> GET /ready HTTP/1.1
> Host: localhost:15000
> User-Agent: curl/7.79.1
> Accept: */*
> 
^C
bash-4.4$ curl http://localhost:15021/healthz/ready -v
*   Trying 127.0.0.1:15021...
* Connected to localhost (127.0.0.1) port 15021 (#0)
> GET /healthz/ready HTTP/1.1
> Host: localhost:15021
> User-Agent: curl/7.79.1
> Accept: */*
> 
^C
bash-4.4$ curl http://localhost:15020/healthz/ready -v
*   Trying 127.0.0.1:15020...
* Connected to localhost (127.0.0.1) port 15020 (#0)
> GET /healthz/ready HTTP/1.1
> Host: localhost:15020
> User-Agent: curl/7.79.1
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Date: Wed, 09 Nov 2022 14:38:33 GMT
< Content-Length: 0
< 
* Connection #0 to host localhost left intact
```
2. I looked a bit into the readiness probe code. From what I understand:
pilot-agent probe to /healthz/ready on **statusPort** (default 15020) https://github.com/istio/istio/blob/089f536c0f26ef7551f0b52a4e812be970ff46fc/pilot/cmd/pilot-agent/status/server.go#L62-L63 will be handled by Probe Check() func https://github.com/istio/istio/blob/089f536c0f26ef7551f0b52a4e812be970ff46fc/pilot/cmd/pilot-agent/status/ready/probe.go#L47 which will eventually reach util.GetUpdateStatusStats() to call envoy admin interface /stats https://github.com/istio/istio/blob/089f536c0f26ef7551f0b52a4e812be970ff46fc/pilot/cmd/pilot-agent/status/util/stats.go#L104-L111 But it doesn't really match the test result on 15020, if probe on 15020 finally reached envoy 15000/stats as in the code, then when we stopped envoy using "kill -STOP PID", probe on 15020 should also fail! 
Also the codes doesn't match my expectation on 15021 because overload will not hinder pilot-agent to reach envoy 15000/stats as in the code, but in overload condition, probe on port 15021 failed. So question is: 
 **1. how should we explain the test results from the codes, or to say what is the request flow of probe on port 15021, does envoy first intercept request before reaching pilot-agent, and then after arriving at agent, it again requests envoy admin /stats?** 2. what is the relation before readinessProbe.httpGet.port value in ingressgateway chart and the statusPort value?https://github.com/istio/istio/blob/089f536c0f26ef7551f0b52a4e812be970ff46fc/manifests/charts/gateways/istio-ingress/templates/deployment.yaml#L137-L141

To sum up, even though readiness probe will not fail on port 15020, it seems this port ONLY check pilot-agent status, but discoupled from envoy process live/readiness status. So changing readiness probe port to 15020 isn't really an appropriate way for us as we will not know if envoy is still ready or not even the probe return 200 OK. But instead, we are wondering if the proposal below would be a possible improvement for readiness probe that istio can make:
in the readiness probe, we make a combination to curl both port 15020 and port 15000 admin to make sure both pilot-agent and envoy process are ready! In this case, we will not have a failed probe also in overload condition as 15000 and 15020 are not stopped by overload actions.
```
exec:
  command:
   - sh
   - -c
   - curl http://localhost:15000/ready
   - curl http://localhost:15020/healthz/ready
```

**Comment by caoyukun0430** [comment]
> This would need envoy side changes to bypass certain listeners/routes/etc from the overload action.
> 
Yukun: Moreover, about this comment I have one doubt. From the code, it seems istio check admin interface if there is updated stats for readiness probe. So it there would be some improvement to bypass certain listeners/routes/etc from the overload action, how would it reflect on /stats so that istio can make use of?

**Comment by howardjohn** [comment]
Istio only checks /stats at startup; once it is ready once it never does again. I would imagine you don't hit overload before ready?

**Comment by caoyukun0430** [comment]
> 
Actually for us, we also configure livenss probe, so that we have real-time check on probes

**Comment by howardjohn** [comment]
I am not sure how that is relevant? Note readiness probes also probe throughout the entire pod lifetime. Istios probe (in our code) just stops hitting /stats after it succeeds once. 

Whether you call that probe manually, through rediness probe, or liveness probe - doesn't matter.

**Comment by caoyukun0430** [comment]
> 15021: envoy -> agent 15020 15020: the actual health check logic
> 
> The agent health checking logic does check that Envoy _started_, but it may miss a case where Envoy were to, for example, deadlock. Istio used to use only the 15020 method for many years
If agent health checking logic does check that Envoy _started_, why it is the case that after we stopped envoy, we could still get 200 OK from 15020? And also in overload condition, 15021 returns 503 but 15020 returns 200OK?

**Comment by howardjohn** [comment]
If you start+stop Envoy, then Envoy _started_. It just later exited. All we check is that the **start** part works; once it starts we stop checking it.

15020 will return 200 since it doesn't check envoy at all (after start), only the agent. 15021 will return 503 since the request goes through the envoy (which is not running)

**Comment by caoyukun0430** [comment]
OK, I see the difference. So back to the problem, changing to port 15020 can't indicate envoy is healthy or not in overload condition because once it started, it always returns 200OK, and 15021 returns 503 which leads to probe failure.
Do you think the combination probe I proposed is an option, or is there any more probe can indicate whether envoy is live/ready in real-time after it started?

**Comment by caoyukun0430** [comment]
After some internal discussion, we would like to ask if it would be reasonable to use curl http://localhost:15000/ready & curl http://localhost:15020/healthz/ready combination for ingressgateway probe to replace the probe to 15021?
The main concern is: would http://localhost:15000/ready be enough to indicate that envoy is healthy? Or to say is there any edge case that envoy is not ready to take traffic but 15000 can still be requested? (other than our overload scenario for sure)

**Comment by howardjohn** [comment]
No, its not, that is why we don't use it. The 15000 port is served by an entirely isolated admin thread that is not related to the data path real requests flow through. It is also prone to timeouts due to XDS spikes.

**Comment by style95** [comment]
@caoyukun0430 
Aside from this issue, I am trying to apply the overload manager to the ingress gateway.
Could you share how did you apply it?
Does this example apply to not only a sidecar but an ingress-gateway?

https://github.com/istio/istio/tree/master/samples/custom-bootstrap

**Comment by leosarra** [comment]
Closing as it is completed by [#52971](https://github.com/istio/istio/pull/52971)
